File: app/src/main/java/me/instabattle/app/activities/LoginActivity.java
package me.instabattle.app.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import me.instabattle.app.R;
import me.instabattle.app.managers.ServiceGenerator;
import me.instabattle.app.managers.UserManager;
import me.instabattle.app.models.Token;
import me.instabattle.app.models.User;
import me.instabattle.app.settings.State;
import me.instabattle.app.utils.Utils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginActivity extends Activity {
    private static String TAG = "LoginActivity";
    private EditText login;
    private EditText password;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        login = (EditText) findViewById(R.id.loginInput);
        password = (EditText) findViewById(R.id.passwordInput);
    }

    public void onSigninClick(View view) {
        final Intent menu = new Intent(this, MenuActivity.class);

        final String username = login.getText().toString();
        final String pass= password.getText().toString();

        Log.d(TAG, "username: " + username + "\t password: " + pass);

        UserManager.getTokenAndDo(username, pass, new Callback<Token>() {
                    @Override
                    public void onResponse(Call<Token> call, Response<Token> response) {
                        if (response.isSuccessful()) {
                            UserManager.getAndDo(username, new Callback<User>() {
                                @Override
                                public void onResponse(Call<User> call, Response<User> response) {
                                    if (response.isSuccessful()) {
                                        State.currentUser = response.body();
                                    } else {
                                        Log.e(TAG, "Something strange");
                                    }
                                }

                                @Override
                                public void onFailure(Call<User> call, Throwable t) {
                                    //TODO
                                    Log.e(TAG, "Cannot GET user");
                                }
                            });
                            State.token = response.body().get();
                            Log.d(TAG, "Got token: " + State.token);
                            ServiceGenerator.initTokenServices();
                            startActivity(menu);
                        } else {
                            Utils.showToast(LoginActivity.this, "Wrong nickname/password combination");
                        }
                    }

                    @Override
                    public void onFailure(Call<Token> call, Throwable t) {
                        Utils.showToast(LoginActivity.this, "Failed to sign in, try again later.");
                        Log.e(TAG, "Cannot obtaining token");
                    }
                });
    }

    public void onSignupClick(View view) {
        Intent signup = new Intent(this, SignupActivity.class);
        startActivity(signup);
    }

    @Override
    public void onBackPressed() {
        // nothing
    }
}

File: app/src/main/java/me/instabattle/app/activities/LoginActivity.kt
package me.instabattle.app.activities

import android.os.Bundle
import android.view.View
import android.widget.EditText
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

import me.instabattle.app.R
import me.instabattle.app.managers.ServiceGenerator
import me.instabattle.app.managers.UserManager
import me.instabattle.app.settings.State
import org.jetbrains.anko.debug
import org.jetbrains.anko.getStackTraceString
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.toast
import retrofit2.HttpException
import java.io.IOException
import java.net.SocketTimeoutException

class LoginActivity : DefaultActivity() {
    private var login: EditText? = null
    private var password: EditText? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        login = findViewById(R.id.loginInput)
        password = findViewById(R.id.passwordInput)
    }

    fun onSignInClick(view: View) {
        val username = login!!.text.toString()
        val pass = password!!.text.toString()

        debug("username: $username\t password: $pass")

        UserManager.getToken(username, pass)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .flatMap {
                    State.token = it.get()
                    debug("Got token: %s".format(State.token))
                    ServiceGenerator.initTokenServices()
                    UserManager.get(username)
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    State.currentUser = it
                    startActivity<MenuActivity>()
                }, {
                    when (it) {
                        is HttpException -> {
                            val errorBody = it.response().errorBody()
                            if (errorBody == null) { // is it even possible?
                                error("Server returned unsuccessful response with empty error and code: %d".format(it.response().code()))
                            } else {
                                //val errorMessage = JSONObject(errorBody.string()).getString("error")
                                toast("Couldn't login")
                                errorBody.close()
                            }
                        }
                        is SocketTimeoutException -> {
                            toast("Request has timed out")
                        }
                        is IOException -> {
                            toast("Unknown network or conversation error")
                            error(it.getStackTraceString())
                        }
                    }
                })
    }

    fun onSignUpClick(view: View) = startActivity<SignUpActivity>()

    override fun onBackPressed() { }
}
